django/template/defaulttags.py

```diff
@@ -1570,3 +1572,94 @@ def do_with(parser, token):
     nodelist = parser.parse(("endwith",))
     parser.delete_first_token()
     return WithNode(None, None, nodelist, extra_context=extra_context)
+
+
+class DefinePartialNode(Node):
+    def __init__(self, partial_name, inline, nodelist):
+        self.partial_name = partial_name
+        self.inline = inline
+        self.nodelist = nodelist
+
+    def render(self, context):
+        return self.nodelist.render(context) if self.inline else ""
+
+
+class RenderPartialNode(Node):
+    def __init__(self, partial_name, partial_mapping):
+        # Defer lookup of nodelist to runtime.
+        self.partial_name = partial_name
+        self.partial_mapping = partial_mapping
+
+    def render(self, context):
+        return self.partial_mapping[self.partial_name].render(context)
+
+
+@register.tag(name="partialdef")
+def partialdef_func(parser, token):
+    """
+    Declare a partial that can be used later in the template.
+
+    Usage::
+
+        {% partialdef partial_name %}
+        Partial content goes here
+        {% endpartialdef %}
+
+    Stores the nodelist in the context under the key "partial_contents" and can
+    be retrieved using the {% partial %} tag.
+
+    The optional ``inline`` argument will render the contents of the partial
+    where it is defined.
+    """
+    tokens = token.split_contents()
+
+    # Check the number of tokens before trying to assign them via indexes.
+    if (tokens_len := len(tokens)) not in (2, 3):
+        name = token.contents.split()[0]
+        raise TemplateSyntaxError(f"{name} tag requires 2-3 arguments")
+
+    partial_name = tokens[1]
+    if tokens_len > 2:
+        inline = tokens[2]
+        if inline != "inline":
+            raise TemplateSyntaxError(
+                "The 'inline' argument does not have any parameters; "
+                "either use 'inline' or remove it completely."
+            )
+    else:
+        inline = False
```

Comment:
Can this be less convoluted by using a match statement?

```suggestion
    match token.split_contents():
        case "partialdef", partial_name, "inline":
            inline = True
        case "partialdef", partial_name, _:
            raise TemplateSyntaxError(
                "The 'inline' argument does not have any parameters; "
                "either use 'inline' or remove it completely."
            )
        case "partialdef", partial_name:
            inline = False
        case _:
            raise TemplateSyntaxError("partialdef tag requires 2-3 arguments")
```

I find this much easier to read.

I realise that this hard-codes the "partialdef" name, but that's something that we can fix if need be - I have ideas of how to tweak the above...

Reply:
> I realise that this hard-codes the "partialdef" name

I don't think that's a problem. The chances of a tag rename are pretty low. 

(template-partials originally had `startpartial`, and `endpartial`, which didn't match Django's `f"end{tokens[0]}"` naming convention, so we updated to `partialdef` and `endpartialdef` but had to parameterise for backwards compatibility with the original names. We don't need that still.)